### ISSUE TITLE
4.12.0 Release checklist

### ISSUE THREAD

**Issue by compulim** [microsoft/BotFramework-WebChat#3731]
# Checklist

## Build

1. [x] Bump MockBot to Bot Framework SDK release 4.12.0
   - https://www.npmjs.com/package/botbuilder/v/4.12.0
   - (We will not bump for patch release)
1. [x] ~Bump `botframework-directlinejs` to `0.14.1`~
   - No newer version
1. [x] Bump to `4.12.0`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.12.0`
   - [x] Run `npm version --no-git-tag-version 4.12.0`
   - [x] Merged into `master`, the PR number is #3739
   - Commit is `63bb97a`
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run daily pipeline manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`218799`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=218799&view=results) and commit is `63bb97a`

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [x] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 88.0.4324.190
   - [x] Edge 88.0.705.81
   - [x] Firefox 86.0
   - [x] IE11 (Windows 10 21322.1000)
   - [x] macOS Safari 13.1.3 (15609.4.1)
      - DLSpeech: Web Socket connection to `westus2.convai.speech.microsoft.com` closed instantly after connected, it also repro on previous versions of Web Chat (failed on 4.9.0, 4.8.1)
         - https://github.com/microsoft/cognitive-services-speech-sdk-js/issues/221
   - [x] iOS Safari 14.4
   - [x] Android Chrome 88.0.4324.181
1. [x] Test specific fixes related to `4.12.0` and previous releases
   
   - [x] Auto scroll snapping
   - [x] Multi-line send box
   - [x] Suggested actions flow layout
   - [x] Suggested actions with multiple lines

## Release
1. [x] Verify on WebChat-release-testing
1. [x] Make sure you are on `master` ~or `qfe`~ branch, run `git status` to check
1. [x] `git pull`
1. [x] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.12.0`
1. [x] `git log`
   - Verify the latest commit is `63bb97a`
1. [x] `git tag v4.12.0`
1. [x] `git push -u upstream v4.12.0`
   - You do not need to kick off a build again, use the previous build
1. [x] Create a new GitHub release, copy entries from `CHANGELOG.md`
   - [x] Subresource Integrity can be generated by
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.12.0/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
      - From local: `cat webchat.js | openssl dgst -sha384 -binary | openssl base64 -A`
   - [x] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -L -o botframework-directlinespeech-sdk-4.12.0.tgz https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.12.0.tgz
         curl -L -o botframework-webchat-4.12.0.tgz https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.12.0.tgz
         curl -L -o botframework-webchat-core-4.12.0.tgz https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.12.0.tgz
         curl -L -o botframework-webchat-api-4.12.0.tgz https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.12.0.tgz
         curl -L -o botframework-webchat-component-4.12.0.tgz https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.12.0.tgz
         ```
1. [x] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [218799](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=218799&view=results) release number is `29` and commit is `63bb97a`
   - Retain the build indefinitely
1. [x] Kick off release to CDN (CDN CUTOFF TIME IS 2PM PST)
   1. [x] Prepare the email for approval
   - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
   - The build number is [`218799`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=218799&view=results), release number is [`34`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?releaseId=2661&_a=release-pipeline-progress) and commit is `63bb97a`
   - Script build number is `131156` (this is fixed)
   1. [x] Send reminder email to approvers
   - Retain the build indefinitely

## Post-release check

- [x] Test using `webchat-release-testing`
   1. [x] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [x] `01.create-react-app`
      1. [x] Nuke `01.create-react-app/node_modules`
      1. [x] `npm install`
      1. [x] `npm install botframework-webchat@4.12.0` (just install the bundle package)
      1. [x] `npm run build`
   1. [x] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.12.0, with subresource integrity
         ```
         <script
            crossorigin="anonymous"
            integrity="sha384-"
            src="https://cdn.botframework.com/botframework-webchat/latest/webchat-es5.js"
         ></script>
         ```
   1. [x] `npx serve` (at repo root)
   1. [x] Go to http://localhost:5000/ to test, including IE11

## Notification to interested parties

- [x] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/370
- [x] Notify related parties for the following fixes
   - [x] SDK team
   - [x] Omnichannel
   - [x] Pooja
- [x] Update root README.md with feature notes
- [x] Update https://github.com/microsoft/BotFramework-sdk/issues/6200
   
# Changelog

### Added

-  Resolves [#2745](https://github.com/microsoft/BotFramework-WebChat/issues/2745). Added new `flow` layout to suggested actions, by [@compulim](https://github.com/compulim) in PR [#3641](https://github.com/microsoft/BotFramework-WebChat/pull/3641) and PR [#3748](https://github.com/microsoft/BotFramework-WebChat/pull/3748)
-  Added new style options to customize auto-scroll, by [@compulim](https://github.com/compulim) in PR [#3653](https://github.com/microsoft/BotFramework-WebChat/pull/3653)
   -  Set `autoScrollSnapOnActivity` to `true` to pause auto-scroll after more than one activity is shown, or a number to pause after X number of activities
   -  Set `autoScrollSnapOnPage` to `true` to pause auto-scroll when a page is filled, or a number between `0` and `1` to pause after % of page is filled
   -  Set `autoScrollSnapOnActivityOffset` and `autoScrollSnapOnPageOffset` to a number (in pixels) to overscroll/underscroll after the pause
-  Supports multiple transcripts in a single composition, by [@compulim](https://github.com/compulim) in PR [#3653](https://github.com/microsoft/BotFramework-WebChat/pull/3653)
-  Resolves [#3368](https://github.com/microsoft/BotFramework-WebChat/issues/3368). Added new `sendBoxButtonAlignment` for button alignment in multi-line text mode, by [@compulim](https://github.com/compulim) in PR [#3668](https://github.com/microsoft/BotFramework-WebChat/pull/3668)
-  Resolves [#3666](https://github.com/microsoft/BotFramework-WebChat/issues/3666). Added support of sovereign clouds when using Direct Line Speech, by [@compulim](https://github.com/compulim) in PR [#3694](https://github.com/microsoft/BotFramework-WebChat/pull/3694)
   -  Please refer to [`DIRECT_LINE_SPEECH.md`](https://github.com/microsoft/BotFramework-WebChat/blob/master/docs/DIRECT_LINE_SPEECH.md#directlinespeechcredentials) for instructions
-  Resolves [#2996](https://github.com/microsoft/BotFramework-WebChat/issues/2996). Added transcript navigation by keyboard navigation keys, by [@compulim](https://github.com/compulim) in PR [#3703](https://github.com/microsoft/BotFramework-WebChat/pull/3703)
-  Resolves [#3544](https://github.com/microsoft/BotFramework-WebChat/issues/3544). Send user ID from props to chat adapter, by [@timenick](https://github.com/timenick) in PR [#3544)(https://github.com/microsoft/BotFramework-WebChat/issues/3544).
-  Resolves [#3562](https://github.com/microsoft/BotFramework-WebChat/issues/3562). Add button word wrap to suggested actions stacked layout, by [@corinagum](https://github.com/corinagum), in PR [#3728](https://github.com/microsoft/BotFramework-WebChat/pull/3728) and [#3745](https://github.com/microsoft/BotFramework-WebChat/pull/3745)
-  Resolves [#3658](https://github.com/microsoft/BotFramework-WebChat/issues/3658). Added new `containerRole` to default style options, by [@nfreear](https://github.com/nfreear) in PR [#3669](https://github.com/microsoft/BotFramework-WebChat/pull/3669)
-  Resolves [#3754](https://github.com/microsoft/BotFramework-WebChat/issues/3754). Added new `useObserveTranscriptFocus` hook, by [@compulim](https://github.com/compulim) in PR [#3755](https://github.com/microsoft/BotFramework-WebChat/pull/3755)
-  Translation for Yue, by [@compulim](https://github.com/compulim) in PR [#3749](https://github.com/microsoft/BotFramework-WebChat/pulls/3749)

### Fixed

-  Fixes [#3278](https://github.com/microsoft/BotFramework-WebChat/issues/3278). Update `HOOKS.md` verbiage, by [@corinagum](https://github.com/corinagum) in PR [#3564](https://github.com/microsoft/BotFramework-WebChat/pull/3564)
-  Fixes [#3534](https://github.com/microsoft/BotFramework-WebChat/issues/3534). Remove 2020 deprecations, by [@corinagum](https://github.com/corinagum) in PR [#3564](https://github.com/microsoft/BotFramework-WebChat/pull/3564) and [#3728](https://github.com/microsoft/BotFramework-WebChat/pull/3728)
-  Fixes [#3561](https://github.com/microsoft/BotFramework-WebChat/issues/3561). Remove MyGet mentions from samples, by [@corinagum](https://github.com/corinagum) in PR [#3564](https://github.com/microsoft/BotFramework-WebChat/pull/3564)
-  Fixes [#3537](https://github.com/microsoft/BotFramework-WebChat/issues/3537). Fix some carousels improperly using aria-roledescription, by [@corinagum](https://github.com/corinagum) in PR [#3599](https://github.com/microsoft/BotFramework-WebChat/pull/3599)
-  Fixes [#3483](https://github.com/microsoft/BotFramework-WebChat/issues/3483). IE11 anchors fixed to open securely without 'noreferrer' or 'noopener', by [@corinagum](https://github.com/corinagum) in PR [#3607](https://github.com/microsoft/BotFramework-WebChat/pull/3607)
-  Fixes [#3565](https://github.com/microsoft/BotFramework-WebChat/issues/3565). Allow strikethrough `<s>` on sanitize markdown, by [@corinagum](https://github.com/corinagum) in PR [#3646](https://github.com/microsoft/BotFramework-WebChat/pull/3646)
-  Fixes [#3672](https://github.com/microsoft/BotFramework-WebChat/issues/3672). Center the icon of send box buttons vertically and horizontally, by [@compulim](https://github.com/compulim) in PR [#3673](https://github.com/microsoft/BotFramework-WebChat/pull/3673)
-  Fixes [#3683](https://github.com/microsoft/BotFramework-WebChat/issues/3683). Activities should be acknowledged when user scrolls to bottom, by [@compulim](https://github.com/compulim) in PR [#3684](https://github.com/microsoft/BotFramework-WebChat/pull/3684)
-  Fixes [#3431](https://github.com/microsoft/BotFramework-WebChat/issues/3431). Race condition between the first bot activity and first user activity should not cause the first bot activity to be delayed, by [@compulim](https://github.com/compulim) in PR [#3705](https://github.com/microsoft/BotFramework-WebChat/pull/3705)
-  Fixes [#3676](https://github.com/microsoft/BotFramework-WebChat/issues/3676). Activities without text should not generate bogus `aria-labelledby`, by [@compulim](https://github.com/compulim) in PR [#3697](https://github.com/microsoft/BotFramework-WebChat/pull/3697)
-  Fixes [#3625](https://github.com/microsoft/BotFramework-WebChat/issues/3625). Update 'no screen reader for custom activity middleware' warning and add screen reader renderer documentation to `ACCESSIBILITY.md`, by [@corinagum](https://github.com/corinagum) in PR [#3689](https://github.com/microsoft/BotFramework-WebChat/pull/3689)
-  Fixes [#3453](https://github.com/microsoft/BotFramework-WebChat/issues/3453). Fixes plain text file attachments to show download link when uploaded, by [@corinagum](https://github.com/corinagum) in PR [#3711](https://github.com/microsoft/BotFramework-WebChat/pull/3711)
-  Fixes [#3612](https://github.com/microsoft/BotFramework-WebChat/issues/3612). Carousel flippers in suggested actions are given extra padding, by [@compulim](https://github.com/compulim) and [@Quirinevwm](https://github.com/Quirinevwm) in PR [#3704](https://github.com/microsoft/BotFramework-WebChat/pull/3704)
-  Fixes [#3411](https://github.com/microsoft/BotFramework-WebChat/issues/3411). With Direct Line Speech, clicking on microphone button during speech recognition should no longer stop working, by [@compulim](https://github.com/compulim) in PR [#3694](https://github.com/microsoft/BotFramework-WebChat/pull/3694)
   -  Although it no locker lock up microphone, clicking on the microphone button has no effect because Direct Line Speech does not support aborting speech recognition
-  Fixes [#3421](https://github.com/microsoft/BotFramework-WebChat/issues/3421). With Direct Line Speech, after not able to recognize any speech, it should no longer stop working, by [@compulim](https://github.com/compulim) in PR [#3694](https://github.com/microsoft/BotFramework-WebChat/pull/3694)
-  Fixes [#3616](https://github.com/microsoft/BotFramework-WebChat/issues/3616). [Accessibility documentation] Update activity timestamp grouping to match visual UI, by [@amal-khalaf](https://github.com/amal-khalaf) in PR [#3708](https://github.com/microsoft/BotFramework-WebChat/pull/3708)
-  Fixes [#3718](https://github.com/microsoft/BotFramework-WebChat/issues/3718). Fixed `webpack.config.js` to use default settings of `['browser', 'module', 'main']` and resolved issues with `uuid` package in IE11, by [@compulim](https://github.com/compulim) in PR [#3726](https://github.com/microsoft/BotFramework-WebChat/pull/3726)
-  Fixes [#3622](https://github.com/microsoft/BotFramework-WebChat/issues/3622). Apply pushed button style options and `aria-pressed` on Adaptive Cards selected buttons, by [@amal-khalaf](https://github.com/amal-khalaf) in PR [#3710](https://github.com/microsoft/BotFramework-WebChat/pull/3710)
-  Fixes [#3618](https://github.com/microsoft/BotFramework-WebChat/issues/3618). Fix Adaptive Cards anchors being disabled when Adaptive Cards is obsolete, by [@corinagum](https://github.com/corinagum) in PR [#3687](https://github.com/microsoft/BotFramework-WebChat/pull/3687)
-  Fixes [#3747](https://github.com/microsoft/BotFramework-WebChat/issues/3747). `aria-pressed` and `aria-role` is not properly set on Adaptive Cards submit buttons, by [@amal-khalaf](https://github.com/amal-khalaf) in PR [#3744](https://github.com/microsoft/BotFramework-WebChat/pull/3744)
-  Fixes [#3750](https://github.com/microsoft/BotFramework-WebChat/issues/3750). Debump Node.js engines requirements for some packages to `12.0.0`, by [@compulim](https://github.com/compulim) in PR [#3753](https://github.com/microsoft/BotFramework-WebChat/pull/3753)
-  Fixes [#3760](https://github.com/microsoft/BotFramework-WebChat/issues/3760). Use `<ErrorBoundary>` to wrap around attachment renderer, by [@compulim](https://github.com/compulim) in PR [#3761](https://github.com/microsoft/BotFramework-WebChat/pull/3761)
-  Fixes [#3764](https://github.com/microsoft/BotFramework-WebChat/issues/3764). Added `role="group"` to the focusable transcript to enable `aria-activedescendant`, by [@compulim](https://github.com/compulim) in PR [#3765](https://github.com/microsoft/BotFramework-WebChat/issues/3765)

### Changed

-  Bumped all dependencies to the latest versions, by [@compulim](https://github.com/compulim) in PR [#3594](https://github.com/microsoft/BotFramework-WebChat/pull/3594), PR [#3694](https://github.com/microsoft/BotFramework-WebChat/pull/3694), and PR [#3544](https://github.com/microsoft/BotFramework-WebChat/pull/3544)
   -  Development dependencies
      -  [`@babel/cli@7.12.1`](https://npmjs.com/package/@babel/cli)
      -  [`@babel/core@7.12.3`](https://npmjs.com/package/@babel/core)
      -  [`@babel/plugin-proposal-class-properties@7.12.1`](https://npmjs.com/package/@babel/plugin-proposal-class-properties)
      -  [`@babel/plugin-proposal-object-rest-spread@7.12.1`](https://npmjs.com/package/@babel/plugin-proposal-object-rest-spread)
      -  [`@babel/plugin-transform-runtime@7.12.1`](https://npmjs.com/package/@babel/plugin-transform-runtime)
      -  [`@babel/preset-env@7.12.1`](https://npmjs.com/package/@babel/preset-env)
      -  [`@babel/preset-react@7.12.5`](https://npmjs.com/package/@babel/preset-react)
      -  [`@babel/preset-typescript@7.12.1`](https://npmjs.com/package/@babel/preset-typescript)
      -  [`@babel/runtime@7.12.5`](https://npmjs.com/package/@babel/runtime)
      -  [`@types/node@14.14.6`](https://npmjs.com/package/@types/node)
      -  [`@types/react@16.9.55`](https://npmjs.com/package/@types/react)
      -  [`@typescript-eslint/eslint-plugin@4.6.1`](https://npmjs.com/package/@typescript-eslint/eslint-plugin)
      -  [`@typescript-eslint/parser@4.6.1`](https://npmjs.com/package/@typescript-eslint/parser)
      -  [`babel-jest@26.6.3`](https://npmjs.com/package/babel-jest)
      -  [`eslint-plugin-prettier@3.1.4`](https://npmjs.com/package/eslint-plugin-prettier)
      -  [`eslint-plugin-react-hooks@4.2.0`](https://npmjs.com/package/eslint-plugin-react-hooks)
      -  [`eslint-plugin-react@7.21.5`](https://npmjs.com/package/eslint-plugin-react)
      -  [`eslint@7.12.1`](https://npmjs.com/package/eslint)
      -  [`husky@4.3.0`](https://npmjs.com/package/husky)
      -  [`jest-image-snapshot@4.2.0`](https://npmjs.com/package/jest-image-snapshot)
      -  [`jest-junit@12.0.0`](https://npmjs.com/package/jest-junit)
      -  [`jest-trx-results-processor@2.2.0`](https://npmjs.com/package/jest-trx-results-processor)
      -  [`jest@26.6.3`](https://npmjs.com/package/jest)
      -  [`lint-staged@10.5.1`](https://npmjs.com/package/lint-staged)
      -  [`lolex@6.0.0`](https://npmjs.com/package/lolex)
      -  [`node-dev@6.2.0`](https://npmjs.com/package/node-dev)
      -  [`node-fetch@2.6.1`](https://npmjs.com/package/node-fetch)
      -  [`prettier@2.1.2`](https://npmjs.com/package/prettier)
      -  [`source-map-loader@1.1.2`](https://npmjs.com/package/source-map-loader)
      -  [`terser-webpack-plugin@4.2.3`](https://npmjs.com/package/terser-webpack-plugin)
      -  [`typescript@4.0.5`](https://npmjs.com/package/typescript)
      -  [`webpack-cli@4.2.0`](https://npmjs.com/package/webpack-cli)
      -  [`webpack-stats-plugin@1.0.2`](https://npmjs.com/package/webpack-stats-plugin)
      -  [`webpack@4.44.2`](https://npmjs.com/package/webpack)
   -  Production dependencies
      -  [`@babel/runtime@7.12.5`](https://npmjs.com/package/@babel/runtime)
      -  [`botframework-directlinejs@0.14.1`](https://npmjs.com/package/botframework-directlinejs)
      -  [`globalize@1.6.0`](https://npmjs.com/package/globalize)
      -  [`markdown-it@12.0.2`](https://npmjs.com/package/markdown-it)
      -  [`microsoft-cognitiveservices-speech-sdk@1.15.1`](https://npmjs.com/package/microsoft-cognitiveservices-speech-sdk)
      -  [`react-redux@7.2.2`](https://npmjs.com/package/react-redux)
      -  [`redux@4.0.5`](https://npmjs.com/package/redux)
      -  [`sanitize-html@2.1.2`](https://npmjs.com/package/sanitize-html)
      -  [`web-speech-cognitive-services@7.1.0`](https://npmjs.com/package/web-speech-cognitive-services)
      -  [`whatwg-fetch@3.4.1`](https://npmjs.com/package/whatwg-fetch)
-  Resolves [#3392](https://github.com/microsoft/BotFramework-WebChat/issues/3392) Bumped Adaptive Cards to the 2.5.0, by [@corinagum](https://github.com/corinagum) in PR [#3630](https://github.com/microsoft/BotFramework-WebChat/pull/3630)

### Samples

-  Fixes [#3473](https://github.com/microsoft/BotFramework-WebChat/issues/3473). Fix samples using activityMiddleware (from 4.10.0 breaking changes), by [@corinagum](https://github.com/corinagum) in PR [#3601](https://github.com/microsoft/BotFramework-WebChat/pull/3601)
-  Fixes [#3582](https://github.com/microsoft/BotFramework-WebChat/issues/3582). Fix Disable Adaptive Cards sample, by [@corinagum](https://github.com/corinagum) in PR [#3687](https://github.com/microsoft/BotFramework-WebChat/pull/3687)
-  Fixes [#3434](https://github.com/microsoft/BotFramework-WebChat/issues/3434). Dispatched event, postBack, or messageBack + activityMiddleware causes fatal error, by [@amal-khalaf](https://github.com/amal-khalaf) in PR [#3671](https://github.com/microsoft/BotFramework-WebChat/pull/3671)
-  Fixes [#3215](https://github.com/microsoft/BotFramework-WebChat/issues/3215). Fix SSO samples `window.opener.postMessage`, by [@corinagum](https://github.com/corinagum) in PR [#3696](https://github.com/microsoft/BotFramework-WebChat/pull/3696)

**Comment by compulim** [comment]
Done.
